build.sh

```diff
@@ -0,0 +1,273 @@
+	DESKTOP_ENVIRONMENT=$([ "$CHOOSED_TYPE" == 'desktop' ] && echo $DESKTOP_ENVIRONMENT || echo '') \
+	DESKTOP_ENVIRONMENT_CONFIG_NAME=$([ "$CHOOSED_TYPE" == 'desktop' ] && echo $DESKTOP_ENVIRONMENT_CONFIG_NAME || echo '') \
+	DESKTOP_APPGROUPS_SELECTED=$([ "$CHOOSED_TYPE" == 'desktop' ] && echo $DESKTOP_APPGROUPS_SELECTED || echo '') \
+	EXPERT=yes \
+	SKIP_EXTERNAL_TOOLCHAINS=yes \
+	CLEAN_LEVEL= \
+	USE_CCACHE=no \
+	COMPRESS_OUTPUTIMAGE=img
+
+if [ "$NOT_CREATE_BURNABLE" = false ];then
+
+echo "Downloading AmlImg..."
+curl -sSL -o ./AmlImg https://github.com/hzyitc/AmlImg/releases/download/$AMLIMG_VERSION/AmlImg_${AMLIMG_VERSION}_linux_amd64
+chmod +x ./AmlImg
+
+echo "Downloading uboot..."
+curl -sSL -o ./uboot.img https://github.com/${UBOOT_REPO}/releases/download/${UBOOT_TAG}/${UBOOT_BURNIMG}
+
+echo "Unpacking uboot..."
+./AmlImg unpack ./uboot.img burn/
+
+echo "Extracting boot and rootfs partitions..."
+diskimg=$(ls output/images/*.img)
+loop=$(sudo losetup --find --show --partscan $diskimg)
+img2simg ${loop}p1 burn/boot.simg
+img2simg ${loop}p2 burn/rootfs.simg
+losetup -d $loop
+
+echo "Generating burnable image..."
+echo -n "sha1sum $(sha1sum burn/boot.simg | awk '{print $1}')" >burn/boot.VERIFY
+echo -n "sha1sum $(sha1sum burn/rootfs.simg | awk '{print $1}')" >burn/rootfs.VERIFY
+cat <<EOF >>burn/commands.txt
+PARTITION:boot:sparse:boot.simg
+VERIFY:boot:normal:boot.VERIFY
+PARTITION:rootfs:sparse:rootfs.simg
+VERIFY:rootfs:normal:rootfs.VERIFY
+EOF
+prefix=$(ls output/images/*.img | sed 's/\.img$//')
+burnimg=${prefix}.burn.img
+./AmlImg pack $burnimg burn/
+fi
+
+echo "Hashing images..."
+
+# There should be only one image if use -n.
+for f in output/images/*.img; do
+	sha256sum "$f" >"${f}.sha"
+	# We don't need to upload them to releases so, no compresses.
+	# xz --threads=0 --compress "$f"
+done
+
+# Image built. now transfer it to the dist directory.
+mv output/images/*.sha $OUTPUT_DIR
+mv output/images/*.img $OUTPUT_DIR
+
+# Now we remove useless caches.
+rm -rf output/ burn/ ./AmlImg *.img *.patch
+echo "You can now find dists at $OUTPUT_DIR"
+echo "All done. You are ready to go :)"
+exit 0
+}
+
+
+# Fetch user options
+OPTIONS=`getopt -o hso:b:r:t:l: --long help,skip-patch,branch:,release:,type:,local-repo-addr:,output-dir: -n $0 -- "$@"`
+if [ $? != 0 ];then
+	usage
+	exit 1
+fi
+eval set -- "${OPTIONS}"
+while true
+do
+    case $1 in
+		-h|--help)
+            usage
+            ;;
+        -b|--branch)
+            CHOOSED_BRANCH=$2
+			[[ ${BRANCH[@]/${CHOOSED_BRANCH}/} != ${BRANCH[@]} ]] && echo "Branch $CHOOSED_BRANCH is choosed." || (echo "You should choose from ${BRANCH[*]}." && exit 1)
+            shift
+            ;;
+        -r|--release)
+            CHOOSED_RELEASE=$2
+			[[ ${RELEASE[@]/${CHOOSED_RELEASE}/} != ${RELEASE[@]} ]] && echo "Release $CHOOSED_RELEASE is choosed." || (echo "You should choose from ${RELEASE[*]}." && exit 1)
+            shift
+            ;;
```

Comment:
All these arguments will be checked by the build framework. So we don't need to check them again.

Reply:
I think this is still necessary because users can immediately know that their parameters are incorrect at the beginning of executing the script, rather than being informed of the error only at compile time.